### ISSUE TITLE
Confusion regarding db read performance after the switch fo fly_postgres_elixir.

### ISSUE THREAD

**Issue by jeissing** [superfly/fly_postgres_elixir#4]
Hello,

I have a test application that is essentially based on Hello-Phoenix (https://github.com/fly-apps/hello-phoenix ).
I initially had this test application including postgres db deployed exclusively in the fra region. Using the classic Repo and running a simple read request (Repo.all(User)), the response time was several hours at an average of 30ms. After I switched the test application to use fly_postgres_elixir and deployed the application including the postgres volumes to the fra and lax regions, the application's read performance deteriorated from 30ms to an average of 300ms. Unfortunately, I cannot explain this performance degradation.

**App-Status**

    vscode ➜ /workspace (main ✗) $ fly status
    App
      Name     = billowing-star-1538          
      Owner    = personal                     
      Version  = 13                           
      Status   = running                      
      Hostname = billowing-star-1538.fly.dev  
    
    Deployment Status
      ID          = 2ab52089-3db7-6cc9-5ef8-77c266ffa7fa         
      Version     = v13                                          
      Status      = successful                                   
      Description = Deployment completed successfully            
      Instances   = 2 desired, 2 placed, 2 healthy, 0 unhealthy  
    
    Instances
    ID       PROCESS VERSION REGION DESIRED STATUS  HEALTH CHECKS      RESTARTS CREATED   
    54bf2e53 app     13      fra    run     running 1 total, 1 passing 0        2h26m ago 
    1b8ff1ed app     13      lax    run     running 1 total, 1 passing 0        2h27m ago 

**Database Status**

  vscode ➜ /workspace (main ✗) $ fly status -a hello-phoenix-database
  
  App
    Name     = hello-phoenix-database          
    Owner    = personal                        
    Version  = 3                               
    Status   = running                         
    Hostname = hello-phoenix-database.fly.dev  
  
  Instances
  ID       PROCESS VERSION REGION DESIRED STATUS            HEALTH CHECKS      RESTARTS CREATED              
  3e6d8469 app     3       lax    run     running (replica) 3 total, 3 passing 0        3h13m ago            
  ebaa9d9a app     3       fra    run     running (replica) 3 total, 3 passing 0        2021-11-01T18:17:40Z 
  ae18bd04 app     3       fra    run     running (leader)  3 total, 3 passing 0        2021-11-01T18:17:30Z 

### log Region fra
2021-11-03T11:47:40.774 app[54bf2e53] fra [info] 11:47:40.774 request_id=FrQG0VRHT3tGIS4AAJkB [info] GET /
2021-11-03T11:47:40.775 app[54bf2e53] fra [info] 11:47:40.775 request_id=FrQG0VRHT3tGIS4AAJkB [info] Sent 200 in 808µs
2021-11-03T11:47:49.533 app[54bf2e53] fra [info] 11:47:49.532 request_id=FrQG015Tpj4V_H8AAJkR [info] GET /users
2021-11-03T11:47:49.838 app[54bf2e53] fra [info] 11:47:49.838 request_id=FrQG015Tpj4V_H8AAJkR [info] Sent 200 in 305ms
2021-11-03T11:47:49.953 app[54bf2e53] fra [info] 11:47:49.953 request_id=FrQG03dit6HuymYAAJkh [info] GET /users
2021-11-03T11:47:50.258 app[54bf2e53] fra [info] 11:47:50.258 request_id=FrQG03dit6HuymYAAJkh [info] Sent 200 in 304ms
2021-11-03T11:47:52.420 app[54bf2e53] fra [info] 11:47:52.419 request_id=FrQG1ApnEQCQShcAAJkx [info] GET /users
2021-11-03T11:47:52.725 app[54bf2e53] fra [info] 11:47:52.725 request_id=FrQG1ApnEQCQShcAAJkx [info] Sent 200 in 305ms
2021-11-03T11:47:54.173 app[54bf2e53] fra [info] 11:47:54.173 request_id=FrQG1HLsVgC4PokAAJlB [info] GET /
2021-11-03T11:47:54.174 app[54bf2e53] fra [info] 11:47:54.174 request_id=FrQG1HLsVgC4PokAAJlB [info] Sent 200 in 871µs

###  Region lax
2021-11-03T11:44:17.573 app[1b8ff1ed] lax [info] 11:44:17.571 request_id=FrQGogRxGvi3RxQAAFhB [info] GET /
2021-11-03T11:44:17.574 app[1b8ff1ed] lax [info] 11:44:17.572 request_id=FrQGogRxGvi3RxQAAFhB [info] Sent 200 in 1ms
2021-11-03T11:44:17.838 app[1b8ff1ed] lax [info] 11:44:17.836 request_id=FrQGohQ_qQBdPLgAAFhR [info] GET /users
2021-11-03T11:44:18.009 app[1b8ff1ed] lax [info] 11:44:18.006 request_id=FrQGohQ_qQBdPLgAAFhR [info] Sent 200 in 170ms

**Comment by brainlid** [comment]
Hi @jeissing! Thanks for trying out the library and giving feedback!

What is your `fly regions list`? And can you show the status of your database app? `fly status -a your-db-app`

**Comment by brainlid** [comment]
Also, since this is a very recently opened issue, I assume you are using version 0.1.9. If not, you should update to that first.

**Comment by jeissing** [comment]
Perhaps also interesting is that the instance in fra uses the following connection strings:

### Primary Fly.Postgres.primary_db_url()
postgres://billowing_star_1538_kzor184w07xemvj5:<password>@fra.hello-phoenix-database.internal:5432/billowing_star_1538?sslmode=disable

### Replica: Fly.Postgres.replica_db_url()
postgres://billowing_star_1538_kzor184w07xemvj5:<password>@fra.hello-phoenix-database.internal:5433/billowing_star_1538?sslmode=disable

**Comment by brainlid** [comment]
Ok. Thanks for the region listing. While this isn't your specific issue, you will hit it eventually. You want to disable backup regions that aren't an intended region.

See [this issue](https://github.com/superfly/fly_postgres_elixir/issues/3#issuecomment-959768752) for details on how to address it. This is the case for multi-region Elixir applications that use postgres replicas (and fly_postgres).


Are you physically closer to fra or lax? I noticed that your primary is in fra.

For the DB connections, that's expected for `fra`. 

You can get an IEx console into the `lax` instance like this:

```
> fly ssh console --select
```
Then use the keyboard to select the instance in `lax`.

I'd be interested to know what you get for the:
```
> Fly.is_primary?
false

> Fly.Postgres.database_url
"postgres://app_95gd38z3loxnwq2v:password@lax.app-db.internal:5433/app?sslmode=disable"
```
The important part here is does it have `@lax.`? It _should_ be connecting to the local replica database.

**Comment by jeissing** [comment]
I am here in Germany and therefore several thousand miles closer to Frankfurt (fra).

**IEx on lax instance**

  iex(billowing-star-1538@fdaa:0:36e6:a7b:85:1b8f:f1ed:2)1> Fly.is_primary?
  false
  
  iex(billowing-star-1538@fdaa:0:36e6:a7b:85:1b8f:f1ed:2)2> Fly.Postgres.database_url
  "postgres://billowing_star_1538_kzor184w07xemvj5:mypassword@hello-phoenix-database.internal:5432/billowing_star_1538?sslmode=disable"
  iex(billowing-star-1538@fdaa:0:36e6:a7b:85:1b8f:f1ed:2)3> 
  21:09:29.617 [info]  Using raw DATABASE_URL. Assumed DEV or TEST environment

**The expected lax prefix is not present.**

**Comment by brainlid** [comment]
That's what I was wondering about. When it's not specified, DNS will just pick one. So you potentially end up connecting to a DB that is far away from you. Then you get the latency you're seeing.

I'm guessing the dockerfile doesn't set the MIX_ENV value the library is expecting. That also needs to be documented!

https://github.com/superfly/fly_postgres_elixir/blob/main/lib/fly_postgres.ex#L54

So it's falling through to use the DATABASE_URL which doesn't specify a region. 

The challenge is dealing with local dev/test environments and production ones.

Just a heads up, I'm currently updating the [hello_elixir repo](https://github.com/fly-apps/hello_elixir) and the Fly Elixir guide for Phoenix 1.6 and other changes. The `hello_elixir` repo has already been updated. You can use it's Dockerfile as an example.

**Comment by jeissing** [comment]
I used the Dockerfile from hello_elixir repository in my project and did a deployment. Without success. In another step, I added MIX_ENV=prod again in the Dockerfile in front of the last line. This time with success. Why the redundant declaration of the environment variable MIX_ENV=prod leads to success is not clear to me. 

.....
RUN set -eux; \
  ln -nfs $(basename *)/bin/$(basename *) /app/entry

**ENV MIX_ENV=prod**
CMD /app/entry start

**Comment by brainlid** [comment]
@jeissing thank you for your multiple experiments to find the fix. Based on your feedback I updated the Dockerfile again. There was a problem with the 2nd setting of the MIX_ENV. There are two Dockefile stages and fly_postgres really needs the 2nd stage one. 

I'm looking into better ways to solve the need so this issue can be avoided. 

Hope that helps!

**Comment by jeissing** [comment]
Yes, now I see it. It is a multistage Dockerfile. The MIX_ENV setting is needed not only at compile time but also at runtime for Fly_postgres_elixir.
Thanks for the quick and professional support. And by the way. I love the massively reduced latency.

**Comment by brainlid** [comment]
@jeissing I released v0.1.10. This includes a CHANGELOG and I updated the README.

This removes the need for the 2nd MIX_ENV setting in the Dockerfile. So you'd still have the first stage `ENV MIX_ENV="prod"` but it doesn't need it for the 2nd release section.

This is a breaking change, in that you need to add a new config for it work properly in production.

In either `config/prod.exs` or `config/runtime.exs` you set the following: 
```
config :fly_postgres, rewrite_db_url: true
```

This works for the production to connect to the primary or replica DB appropriately. In dev and test it leave the DB config alone. Give it a try! Hope this cleans up some of that for you!

**Comment by jeissing** [comment]
I have migrated to v.0.1.10. It works as expected. Thank you very much.
